src/remote.c

```diff
@@ -51,62 +54,66 @@ uint64_t remotehston(const uint32_t max, const char *const str)
 }
 
 #if PC_HOSTED == 0
-static void remote_send_buf(uint8_t *buffer, size_t len)
+/* hex-ify and send a buffer of data */
+static void remote_send_buf(const void *const buffer, const size_t len)
 {
-	uint8_t *p = buffer;
-	char hex[2];
-	do {
-		hexify(hex, (const void *)p++, 1);
-
+	char hex[2] = {};
+	const uint8_t *const data = (const uint8_t *)buffer;
+	for (size_t offset = 0; offset < len; ++offset) {
+		hexify(hex, data + offset, 1U);
 		gdb_if_putchar(hex[0], 0);
 		gdb_if_putchar(hex[1], 0);
-
-	} while (p < (buffer + len));
+	}
 }
 
-static void remote_respond_buf(char respCode, uint8_t *buffer, size_t len)
+/* Send a response with some data following */
+static void remote_respond_buf(const char response_code, const void *const buffer, const size_t len)
 {
 	gdb_if_putchar(REMOTE_RESP, 0);
-	gdb_if_putchar(respCode, 0);
+	gdb_if_putchar(response_code, 0);
 
 	remote_send_buf(buffer, len);
 
 	gdb_if_putchar(REMOTE_EOM, 1);
 }
 
-/* Send response to far end */
-static void remote_respond(char respCode, uint64_t param)
+/* Send a response with a simple result code parameter */
+static void remote_respond(const char response_code, uint64_t param)
 {
-	char buf[35]; /*Response, code, EOM and 2*16 hex nibbles*/
-	char *p = buf;
+	/* Put out the start of response marker and response code */
+	gdb_if_putchar(REMOTE_RESP, false);
+	gdb_if_putchar(response_code, false);
 
-	gdb_if_putchar(REMOTE_RESP, 0);
-	gdb_if_putchar(respCode, 0);
+	/* This defines space for exactly a 64-bit number expressed in hex */
+	char response[16];
+	size_t idx = 0;
 
-	do {
```

Comment:
The original do-while loop produced a different protocol: there was always at least one hex digit in a successful result. Unfortunately, the consuming code in BMDA (`remote_swdptap.c`) wasn't adjusted, so there were spurious errors whenever an operation produced an all-zeros response. (`K` instead of `K0` like the previous version did)

Reply:
We are split between adjusting this and changing the remote_swdptap.c code - there are a couple of ways we can adjust the new loop to regain the old behaviour but at the cost of code size; it's probably better to adjust BMDA instead

Reply:
Hmm.. this is actually screwing inadvertently with the old JTAG scan method too because it makes jtag_proc.jtagtap_next() produce different results to what's expected.. we'll check if idx is 0 after the loop and adjust it in a fixup step. That'll give the old protocol behaviour

Reply:
As long as BMDA still accepts `K0`, it'll probably remain compatible with older firmware. I prefer to avoid changing existing protocols except where absolutely necessary, lest there be some subtle compatibility issue later on.

(The existing `remote_swdptap.c` also has an incorrect check in the debug message that caused a `swdptap_seq_in_parity failed, error \n` instead of usefully saying `short response`.)

My thoughts about how to adjust the new loop are to either change the condition to `idx < 16 && (param || !idx)` (which is what I tested) or just move the non-zero check for `param` to after the shift, and have it break the loop

```c
    param >>= 4U;
    if (!param)
        break;
```

I haven't inspected the assembly output to compare them, but it's worth noting that the existing proposed change probably already has two test-and-branch sections in the loop already, which is probably more than the previous code did.

Reply:
Interestingly, the new code produces overall smaller asm output than the old because the compiler is more readily able to optimise. In this case, we do get to entirely change the protocol if we wish as we bumped the protocol version for that reason - a license to break compatibility - but fixing `K` meaning `K0` in every remote protocol function in BMDA would be far too much work. Please give the current version of this PR a spin as it should have fixed all 3 of the issues you found :)